05_stringer/runnerdave/main_test.go

```diff
@@ -0,0 +1,49 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote("127.0.0.1\n")
+	actual := strconv.Quote(buf.String())
+	t.Logf("expected:%s", expected)
+	t.Logf("actual:%s", actual)
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main()")
+	}
+}
+
+func TestStringer(t *testing.T) {
+	tables := []struct {
+		x IPAddr
+		n string
+	}{
+		{IPAddr{127, 0, 0, 1}, "127.0.0.1"},
+		{IPAddr{10, 0, 0, 1}, "10.0.0.1"},
+		{IPAddr{191, 255, 255, 255}, "191.255.255.255"},
+		{IPAddr{10, 10}, "10.10.0.0"},
+		{IPAddr{}, "0.0.0.0"},
+	}
+
+	for _, table := range tables {
+		var buf bytes.Buffer
+		out = &buf
+		fmt.Fprintln(out, table.x)
+		expected := strconv.Quote(table.n)
+		actual := strconv.Quote(strings.Replace(buf.String(), "\n", "", -1))
```

Comment:
Using `Fprintln` to call the `String` method and then stripping off the `\n` added by `Fprintln` seems unnecessary. You could just call `Fprint` instead (https://golang.org/pkg/fmt/#Fprint) and not have a newline to strip off.

I'd also consider removing `Fprint(ln)` altogether and just call `table.x.String()`. Using a buffer to capture the output is required for `main`, but when testing a function with inputs and outputs, it makes more sense to deal directly with those inputs and outputs with nothing in between.

Finally, since there are no special characters in the string literals now, there is no need to use `strconv.Quote`.

With all this in place, your test should just be:
```
expected := table.n
actual := table.x.String()
```

That is much simpler to read an understand.

On the subject of reading and understanding, using `n` and `x` for the fields in your struct are rather meaningless and even misleading. `n` is usually some sort of count (int). I think naming them something like `input` (or `in`) and `expected` (or `out`) would make this more readable. Go encourages short variable names, but they need to be meaningful. In this context, `n` and `x` are not meaningful. They just seem to be random characters.

If you are unsure of naming, it is often useful to look at the Go library source code. E.g. https://golang.org/src/strconv/atob_test.go (the strconv package) uses `in` and `out`, so following that means you likely wont go wrong.

Reply:
@camh-anz  this is really valuable feedback thanks a lot, will implement later